FWCore/Framework/src/EventProcessor.cc

```diff
@@ -1105,377 +1102,663 @@ namespace edm {
+                          input_->doBeginLumi(lumiPrincipal, &processContext_);
+                          sentry.completedSuccessfully();
                         }
-                      }
-                    }) | runLast(postQueueTask);
 
-                  } catch (...) {
-                    status->resetResources();
-                    postQueueTask.doneWaiting(std::current_exception());
-                  }
-                });  // task in sourceResourcesAcquirer
-          });
+                        Service<RandomNumberGenerator> rng;
+                        if (rng.isAvailable()) {
+                          LuminosityBlock lb(lumiPrincipal, ModuleDescription(), nullptr, false);
+                          rng->preBeginLumi(lb);
+                        }
+
+                        EventSetupImpl const& es = status->eventSetupImpl(esp_->subProcessIndex());
+
+                        using namespace edm::waiting_task::chain;
+                        chain::first([this, status](auto nextTask) mutable {
+                          readAndMergeLumiEntries(std::move(status), std::move(nextTask));
+                          firstItemAfterLumiMerge_ = true;
+                        }) | then([this, status, &es, &lumiPrincipal](auto nextTask) {
+                          LumiTransitionInfo transitionInfo(lumiPrincipal, es, &status->eventSetupImpls());
+                          using Traits = OccurrenceTraits<LuminosityBlockPrincipal, BranchActionGlobalBegin>;
+                          beginGlobalTransitionAsync<Traits>(
+                              nextTask, *schedule_, transitionInfo, serviceToken_, subProcesses_);
+                        }) | ifThen(looper_, [this, status, &es](auto nextTask) {
+                          looper_->prefetchAsync(
+                              nextTask, serviceToken_, Transition::BeginLuminosityBlock, *(status->lumiPrincipal()), es);
+                        }) | ifThen(looper_, [this, status, &es](auto nextTask) {
+                          status->globalBeginDidSucceed();
+                          ServiceRegistry::Operate operateLooper(serviceToken_);
+                          looper_->doBeginLuminosityBlock(*(status->lumiPrincipal()), es, &processContext_);
+                        }) | then([this, status, iRunStatus](std::exception_ptr const* iException, auto holder) mutable {
+                          if (iException) {
+                            status->resetResources();
+                            queueWhichWaitsForIOVsToFinish_.resume();
+                            WaitingTaskHolder copyHolder(holder);
+                            copyHolder.doneWaiting(*iException);
+                            endRunAsync(iRunStatus, holder);
+                          } else {
+                            if (not looper_) {
+                              status->globalBeginDidSucceed();
+                            }
+
+                            status->setGlobalEndRunHolder();
```

Comment:
Here is the only place the lumi version of globalEndRunHolderDoneWaiting is called. At this point, we have access to iRunStatus and therefore we could just pass the WaitingTaskHolder at this point.

Reply:
See comment above.